driver-core/src/main/com/mongodb/client/model/expressions/DocumentExpression.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client.model.expressions;
+
+import org.bson.conversions.Bson;
+import org.bson.types.Decimal128;
+
+import java.time.Instant;
+
+import static com.mongodb.client.model.expressions.Expressions.of;
+
+/**
+ * Expresses a document value. A document is an ordered set of fields, where the
+ * key is a string value, mapping to a value of any other expression type.
+ */
+public interface DocumentExpression extends Expression {
+
+    DocumentExpression setField(String fieldName, Expression exp);
+
+    DocumentExpression unsetField(String fieldName);
+
+    Expression getField(String fieldName);
+
+    BooleanExpression getBoolean(String fieldName);
+
+    BooleanExpression getBoolean(String fieldName, BooleanExpression other);
+
+    default BooleanExpression getBoolean(final String fieldName, final boolean other) {
+        return getBoolean(fieldName, of(other));
+    }
+
+    NumberExpression getNumber(String fieldName);
+
+    NumberExpression getNumber(String fieldName, NumberExpression other);
+
+    default NumberExpression getNumber(final String fieldName, final double other) {
+        return getNumber(fieldName, of(other));
+    }
+
+    default NumberExpression getNumber(final String fieldName, final Decimal128 other) {
```

Comment:
@stIncMale / see [prior discussion](https://github.com/mongodb/mongo-java-driver/pull/1052#discussion_r1031944153)

I plan to revert these methods back to Number, because of the following difference between of and getT:

```
        assertExpression(99, ofDoc("{}").getInteger("a", 99));
        assertExpression(99, ofDoc("{}").getNumber("a", 99)); // fails round-trip
        assertExpression(99L, ofDoc("{}").getNumber("a", 99L)); // fails round-trip
        assertExpression(99.0, ofDoc("{}").getNumber("a", 99.0));
        assertExpression(99, of(99));
        assertExpression(99L, of(99L));
        assertExpression(99.0, of(99.0));
        assertExpression(99, ofDoc("{}").getNumber("a", of(99)));
        assertExpression(99L, ofDoc("{}").getNumber("a", of(99L)));
```
If we use Number on `of`, then we must have both `ofInteger` and `ofNumber`. If we provide both a `getInteger(int)` and a `getNumber(int)`, then our intention in providing both methods seems confusing to users (if we did not catch the above round-tripping failure, they will not think it obvious).

Reply:
For clarity, the "round-tripping" is perhaps better described as a general problem of user expectations about types. It is true that technically, the failing `.getNumber("a", 99)` accepts a double rather than an int, but that is irrelevant since the user sees what appears to be an int, and would expect it to be accepted as an int, and then come back as an int.

Reply:
Reverted in https://github.com/mongodb/mongo-java-driver/pull/1053#discussion_r1060769700